filters/HAGFilter.cpp

```diff
@@ -229,6 +242,27 @@ void HAGFilter::filter(PointView& view)
 {
     using namespace pdal::Dimension;
 
+    // If using a raster as the DTM, check each point against it and write
```

Comment:
This should go in a function.

This behavior is somewhat different than the other methods: using other methods we always set HAG for points classified as ground to 0.  This doesn't guarantee that.  @chambbj , do you have thoughts on this?

Reply:
It is true that the original method, which relied on ground classifications, would set HAG equal to 0 for these ground classified points. For the rest, it was valid to have both positive and negative HAG values.

In this case, I suppose you could apply some logic that if a point was classified as zero and was sufficiently close to the DEM, then you could force the HAG to zero. That said, with the external DEM, if you are treating it as truth, maybe we don't even require ground classifications as a prerequisite anymore, and just let the HAG calculation carry through as written.

Reply:
I don't disagree, but I have two concerns:

1) The behavior is inconsistent.
2) You will probably create points classified as ground that have a non-zero HeightAboveGround.

Not sure how important those concerns are.

Reply:
We can think about how to make it more consistent.

To your second point, I guess I'm suggesting that points don't have to be classified as anything first, but you are right, if they are, they could be non-zero.

I could envision a number of possibilities to address the second point. One, flip things around and use HAG directly to classify ground points (if abs(HAG) < threshold, set Classification to ground). Two, use Classification to modify HAG (if Classification = ground, set HAG to 0). Both of these are reasonable post-processing steps that we don't really have a filter to do (outside of Python). They could be baked in here, or we could wait on #2638.

Reply:
I get a little concerned about complication that might be involved in the above suggestion.  Options that change significant behavior for a filter don't get me excited.

I'd rather just see things operate the same way they do currently -- if the point is classified as ground, set HAG to 0.  If not, use the raster value.  There is also a question of what to do if there is no raster cell where a point falls.

Classifying points based on a raster seems more like something for `filters.dem` or a new filter.

Reply:
Right, but `filters.dem` culls points. Even more confusing to stick it there. New filter, maybe.

Perhaps the simplest thing to do right now would be to add an option to force HAG to 0 if Classifications exist and it is ground - default would be ON. I do think that we should allow the filter to proceed without Classifications at all if the DEM option is exercised.

Reply:
I've implemented what you suggested @chambbj - I've used `respect_ground_classification` as the option but don't know if its the best choice. Open to suggestions here.

Reply:
I think this is the right way to approach this. A more succinct option name would be good. I'm at a bit of a loss, maybe `zero_ground`?